pkg/cvo/updatepayload.go

```diff
@@ -314,28 +314,102 @@ func (r *payloadRetriever) pruneJobs(ctx context.Context, retain int) error {
+		matches = append(matches, current)
+		if equalDigest(update.Image, current.Image) {
+			foundImage = true
+		}
 	}
-	return *update, true
+
+	for _, release := range config.Status.AvailableUpdates {
+		if warning, err := releaseMatchesUpdate(&release, update); err != nil {
+			klog.V(2).Infof("Excluding recommended update %s from matching: %s", release.Version, err)
+			continue
+		} else if warning != nil {
+			mismatchedVersionWarning = warning
+		}
+		matches = append(matches, release)
+		if equalDigest(update.Image, release.Image) {
+			foundImage = true
+		}
+	}
+
+	if len(update.Image) > 0 && !foundImage {
+		// extend the set of candidates with the spec's explicit image
+		matches = append(matches, configv1.Release{
+			Version: update.Version,
+			Image:   update.Image,
+		})
+	}
+
+	if len(matches) == 0 {
+		return currentUpdate, fmt.Errorf("no available updates found matching %v", update)
+	}
+
+	if len(matches) > 1 {
+		if mismatchedVersionWarning != nil {
+			return currentUpdate, fmt.Errorf("multiple available updates found matching %v: %w", update, mismatchedVersionWarning)
+		}
+		return currentUpdate, fmt.Errorf("multiple available updates found matching %v", update)
+	}
+
+	// Single match happy case.  Fill in any properties in the request that were left implicit
+
+	desired := *update
+	if desired.Version == "" {
+		desired.Version = matches[0].Version
+	}
+	if desired.Image == "" {
+		desired.Image = matches[0].Image
+	} else if mismatchedVersionWarning != nil {
+		desired.Version = "" // avoid tripping VerifyPayloadVersionFailed.  TODO: eventually drop this, if we can demonstrate that incompatible (image, version) pairs are uncommon in the wild
+	}
+
+	return desired, mismatchedVersionWarning
 }
 
-func findUpdateFromConfigVersion(config *configv1.ClusterVersion, version string, force bool) (configv1.Update, bool) {
-	for _, update := range config.Status.AvailableUpdates {
-		if update.Version == version && len(update.Image) > 0 {
-			return configv1.Update{
-				Version: version,
-				Image:   update.Image,
-				Force:   force,
-			}, true
+// releaseMatchesUpdate returns true if and only if the release
+// matches the update filters.  Includes both warning and error modes,
+// to allow backwards compatibilty with updates that specify a release
+// image and an incompatible version.
+func releaseMatchesUpdate(release *configv1.Release, update *configv1.Update) (error, error) {
+	if release == nil || release.Image == "" {
+		return nil, fmt.Errorf("releases without images never match: %#v", release)
+	}
+
+	var mismatchedVersionError error
+	if update.Version != "" && release.Version != update.Version {
+		mismatchedVersionError := fmt.Errorf("release version %q does not match desired update %q", release.Version, update.Version)
+		if mismatchedVersionError != nil && update.Image == "" { // TODO: eventually drop the image-check here, if we can demonstrate that incompatible (image, version) pairs are uncommon in the wild
```

Comment:
`mismatchedVersionError != nil` check is weird here, it's never nil, it's always set on the line just above?

Reply:
Dropped in 6c2d7e79 -> 1618a424